### PR TITLE
Set the HasExplicitSize flag when a class size is specified in metadata.

### LINKED ISSUE #24705: WPF Applications crashing when loading DWrite
* .NET Core Version: 3.0.100-preview6-012012, regression from preview5, exact build that regresses unknown at this time.
* Windows version: N/A
* Does the bug reproduce also in WPF for .NET Framework 4.8?:No

 
 
 **Problem description:**
 When a WPF application loads DirectWriteForwarder, the module constructor attempts to load DirectWrite and instantiate the appropriate factory for use with font rendering.  In this build, the invocation of the creation function for the factory fails with an invalid cast HResult.

The offending function is in Factory,cpp on line 122 (pasted here as this source is not currently available):
```c#
  __declspec(noinline) void Factory::Initialize(
                                                 FactoryType factoryType
                                                 )
    {
        IUnknown* factoryTemp;
        DWRITECREATEFACTORY pfnDWriteCreateFactory = (DWRITECREATEFACTORY)GetDWriteCreateFactoryFunctionPointer();

        HRESULT hr = (*pfnDWriteCreateFactory)(
            DWriteTypeConverter::Convert(factoryType),
            (REFIID)(*(_guidForIDWriteFactory->Value)),
            &factoryTemp
            );
        
        ConvertHresultToException(hr, "Factory::Initialize");

        _pFactory = (IDWriteFactory*)factoryTemp;
    }
```
 **Actual behavior:** Invalid cast exception.
 
 **Expected behavior:** DirectWrite is initialized.
 
 **Minimal repro:** https://github.com/rladuca/DWriteIssue

@dotnet/wpf-developers

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Fixes #24705

### CODE REVIEW

**Comment by jkoritzinsky** [comment]
I'll try to add a test case for it.

File: tests/src/Loader/classloader/SequentialLayout/Regress/ExplicitSize/ExplicitSize.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+// The struct has an objref and is of odd size.
+// The GC requires that all valuetypes containing objrefs be sized to a multiple of sizeof(void*) )== 4).
+// Since the size of this struct was 17 we were throwing a TypeLoadException.
```
Comment (mikedn) [inline tests/src/Loader/classloader/SequentialLayout/Regress/ExplicitSize/ExplicitSize.cs]: Seems like this comment does not belong here.

File: src/vm/class.cpp [hunk 2]
```diff
@@ -3767,6 +3767,10 @@ VOID EEClassLayoutInfo::CollectLayoutFieldMetadataThrowing(
     {
         classSizeInMetadata = 0;
     }
+    else
```
Comment (AaronRobinsonMSFT) [inline src/vm/class.cpp]: Please comment on the logic here. If we didn't fail to get the size, this indicates we have an explicit layout? Which would seem to imply we can check `cl` much earlier.

File: tests/src/Loader/classloader/SequentialLayout/Regress/ExplicitSize/ExplicitSize.cs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+{
+    [StructLayout(LayoutKind.Sequential, Size = 16)]
+    struct GUID
+    {
+        private int align;
+    }
+
+    static int Main()
+    {
+        Guid initialGuid = Guid.NewGuid();
```
[the remark this answers is not in the document]
Comment (mikedn) [reply tests/src/Loader/classloader/SequentialLayout/Regress/ExplicitSize/ExplicitSize.cs]: Not a fan of having to write stuff to the console in tests, I think hardcoding a `Guid` here would be fine. The bug manifests by copying only a part of the `Guid` so any non-empty `Guid` will do.

File: tests/src/Loader/classloader/SequentialLayout/Regress/ExplicitSize/ExplicitSize.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply tests/src/Loader/classloader/SequentialLayout/Regress/ExplicitSize/ExplicitSize.cs]: That's a copy/paste issue on my end. I'll remove that.

File: src/vm/class.cpp (hunk 2 above)
Comment (jkoritzinsky) [reply src/vm/class.cpp]: It's not explicit layout, it's explicitly defined size (like `[StructLayout(..., Size = N)]`. I'll add a comment.

**Pull request merged**